File: sdks/python/clawdb/provision.py
from __future__ import annotations

import asyncio
import os
import shutil
import socket
import subprocess
import tarfile
from dataclasses import dataclass
from pathlib import Path

import httpx


@dataclass(frozen=True)
class ProvisionResult:
    endpoint: str
    api_key: str | None


def _is_port_open(host: str, port: int, timeout: float = 0.2) -> bool:
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
        sock.settimeout(timeout)
        return sock.connect_ex((host, port)) == 0


def _platform_id() -> str:
    import platform
    import sys

    machine = platform.machine().lower()
    if sys.platform == 'linux' and machine in ('x86_64', 'amd64'):
        return 'linux-x64'
    if sys.platform == 'linux' and machine in ('aarch64', 'arm64'):
        return 'linux-arm64'
    if sys.platform == 'darwin' and machine in ('x86_64', 'amd64'):
        return 'darwin-x64'
    if sys.platform == 'darwin' and machine in ('arm64', 'aarch64'):
        return 'darwin-arm64'
    if sys.platform == 'win32' and machine in ('x86_64', 'amd64'):
        return 'win32-x64'
    raise RuntimeError(f'Unsupported platform: {sys.platform}-{machine}')


def _download_binary() -> Path:
    platform_id = _platform_id()
    bin_dir = Path.home() / '.clawdb' / 'bin'
    bin_dir.mkdir(parents=True, exist_ok=True)

    archive_name = f'clawdb-server-{platform_id}.tar.gz'
    archive_path = bin_dir / archive_name

    for base in _release_base_candidates():
        with httpx.Client(timeout=10.0, follow_redirects=True) as client:
            archive_resp = client.get(f'{base}/{archive_name}')
            if archive_resp.status_code >= 400:
                continue
            archive_path.write_bytes(archive_resp.content)
            return archive_path

    raise RuntimeError('Unable to locate downloadable clawdb-server release archive')


def _release_base_candidates() -> list[str]:
    configured = os.getenv('CLAWDB_SERVER_RELEASE_BASE_URL', '').strip()
    candidates = [
        configured,
        'https://github.com/Claw-DB/ClawDB/releases/latest/download',
        'https://github.com/clawdb/clawdb/releases/latest/download',
        'https://github.com/claw-db/clawdb/releases/latest/download',
    ]
    return list(dict.fromkeys([candidate for candidate in candidates if candidate]))


def _binary_name() -> str:
    return 'clawdb-server.exe' if os.name == 'nt' else 'clawdb-server'


def _installed_binary_path() -> Path:
    return Path.home() / '.clawdb' / 'bin' / _binary_name()


def _path_binary() -> str | None:
    return shutil.which('clawdb-server')


def _extract_binary(archive_path: Path) -> Path:
    bin_dir = archive_path.parent
    final_binary = _installed_binary_path()

    with tarfile.open(archive_path, 'r:gz') as archive:
        archive.extractall(bin_dir)

    candidates = [
        bin_dir / f'clawdb-server-{_platform_id()}',
        bin_dir / 'clawdb-server',
        bin_dir / 'clawdb-server.exe',
    ]
    extracted = next((candidate for candidate in candidates if candidate.exists()), None)
    if extracted is None:
        raise RuntimeError('Downloaded archive did not contain clawdb-server binary')

    if extracted != final_binary:
        extracted.replace(final_binary)
    if os.name != 'nt':
        final_binary.chmod(0o755)
    return final_binary


def _spawn_local_server(binary: str | Path) -> None:
    env = os.environ.copy()
    env.setdefault('CLAW_GUARD_JWT_SECRET', 'clawdb-sdk-local-dev-secret')
    env.setdefault('CLAW_VECTOR_ENABLED', 'false')
    subprocess.Popen(
        [str(binary), '--grpc-port', '50050'],
        env=env,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
        start_new_session=True,
    )


async def _wait_for_local_server(timeout_s: float = 5.0) -> bool:
    deadline = asyncio.get_running_loop().time() + timeout_s
    while asyncio.get_running_loop().time() < deadline:
        if _is_port_open('127.0.0.1', 50050):
            return True
        await asyncio.sleep(0.1)
    return False


async def resolve_endpoint() -> ProvisionResult:
    explicit_url = os.getenv('CLAWDB_URL')
    if explicit_url:
        return ProvisionResult(endpoint=explicit_url, api_key=None)

    api_key = os.getenv('CLAWDB_API_KEY')
    if api_key:
        return ProvisionResult(endpoint='https://cloud.clawdb.dev', api_key=api_key)

    if _is_port_open('127.0.0.1', 50050):
        return ProvisionResult(endpoint='http://localhost:50050', api_key=None)

    try:
        path_binary = _path_binary()
        if path_binary:
            _spawn_local_server(path_binary)
            if await _wait_for_local_server():
                return ProvisionResult(endpoint='http://localhost:50050', api_key=None)

        installed_binary = _installed_binary_path()
        if installed_binary.exists():
            _spawn_local_server(installed_binary)
            if await _wait_for_local_server():
                return ProvisionResult(endpoint='http://localhost:50050', api_key=None)

        archive_path = _download_binary()
        extracted_binary = _extract_binary(archive_path)
        _spawn_local_server(extracted_binary)
        if await _wait_for_local_server():
            return ProvisionResult(endpoint='http://localhost:50050', api_key=None)
    except Exception as exc:
        raise RuntimeError(
            'Unable to auto-provision local clawdb-server. Start clawdb-server manually, set CLAWDB_URL, or set CLAWDB_API_KEY.'
        ) from exc

    raise RuntimeError(
        'Unable to auto-provision local clawdb-server. A local server did not become healthy after startup.'
    )

File: sdks/go/provision.go
package clawdb

import (
	"archive/tar"
	"compress/gzip"
	"fmt"
	"io"
	"net"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

func releaseBaseCandidates() []string {
	configured := strings.TrimSpace(os.Getenv("CLAWDB_SERVER_RELEASE_BASE_URL"))
	candidates := []string{
		configured,
		"https://github.com/Claw-DB/ClawDB/releases/latest/download",
		"https://github.com/clawdb/clawdb/releases/latest/download",
		"https://github.com/claw-db/clawdb/releases/latest/download",
	}
	result := make([]string, 0, len(candidates))
	seen := map[string]struct{}{}
	for _, candidate := range candidates {
		if candidate == "" {
			continue
		}
		if _, ok := seen[candidate]; ok {
			continue
		}
		seen[candidate] = struct{}{}
		result = append(result, candidate)
	}
	return result
}

// Options configures the Go client using a single struct.
type Options struct {
	Endpoint  string
	APIKey    string
	AgentID   string
	Workspace string
	Role      string
	Timeout   time.Duration
	TLS       bool
}

func applyOptionsStruct(cfg *Config, options Options) {
	if options.Endpoint != "" {
		cfg.Endpoint = options.Endpoint
	}
	if options.APIKey != "" {
		cfg.APIKey = options.APIKey
	}
	if options.AgentID != "" {
		cfg.AgentID = options.AgentID
	}
	if options.Workspace != "" {
		cfg.Workspace = options.Workspace
	}
	if options.Role != "" {
		cfg.Role = options.Role
	}
	if options.Timeout > 0 {
		cfg.Timeout = options.Timeout
	}
	if options.TLS {
		cfg.TLS = true
	}
}

func platformID() (string, error) {
	switch runtime.GOOS + "/" + runtime.GOARCH {
	case "linux/amd64":
		return "linux-x64", nil
	case "linux/arm64":
		return "linux-arm64", nil
	case "darwin/amd64":
		return "darwin-x64", nil
	case "darwin/arm64":
		return "darwin-arm64", nil
	case "windows/amd64":
		return "win32-x64", nil
	default:
		return "", fmt.Errorf("unsupported platform: %s/%s", runtime.GOOS, runtime.GOARCH)
	}
}

func localServerHealthy() bool {
	conn, err := net.DialTimeout("tcp", "127.0.0.1:50050", 300*time.Millisecond)
	if err != nil {
		return false
	}
	_ = conn.Close()
	return true
}

func spawnLocalServer(binary string) error {
	cmd := exec.Command(binary, "--grpc-port", "50050")
	cmd.Env = os.Environ()
	if os.Getenv("CLAW_GUARD_JWT_SECRET") == "" {
		cmd.Env = append(cmd.Env, "CLAW_GUARD_JWT_SECRET=clawdb-sdk-local-dev-secret")
	}
	if os.Getenv("CLAW_VECTOR_ENABLED") == "" {
		cmd.Env = append(cmd.Env, "CLAW_VECTOR_ENABLED=false")
	}
	cmd.Stdout = io.Discard
	cmd.Stderr = io.Discard
	return cmd.Start()
}

func waitForLocalServer(timeout time.Duration) bool {
	deadline := time.Now().Add(timeout)
	for time.Now().Before(deadline) {
		if localServerHealthy() {
			return true
		}
		time.Sleep(100 * time.Millisecond)
	}
	return false
}

func binaryName() string {
	if runtime.GOOS == "windows" {
		return "clawdb-server.exe"
	}
	return "clawdb-server"
}

func installedBinaryPath() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	return filepath.Join(home, ".clawdb", "bin", binaryName()), nil
}

func extractBinary(archivePath string) (string, error) {
	archiveFile, err := os.Open(archivePath)
	if err != nil {
		return "", err
	}
	defer archiveFile.Close()

	gzipReader, err := gzip.NewReader(archiveFile)
	if err != nil {
		return "", err
	}
	defer gzipReader.Close()

	tarReader := tar.NewReader(gzipReader)
	binDir := filepath.Dir(archivePath)
	finalBinary, err := installedBinaryPath()
	if err != nil {
		return "", err
	}

	for {
		header, err := tarReader.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return "", err
		}
		if header.Typeflag != tar.TypeReg {
			continue
		}
		name := filepath.Base(header.Name)
		if name != "clawdb-server" && name != "clawdb-server.exe" && !strings.HasPrefix(name, "clawdb-server-") {
			continue
		}

		out, err := os.Create(finalBinary)
		if err != nil {
			return "", err
		}
		if _, err := io.Copy(out, tarReader); err != nil {
			_ = out.Close()
			return "", err
		}
		if err := out.Close(); err != nil {
			return "", err
		}
		if runtime.GOOS != "windows" {
			if err := os.Chmod(finalBinary, 0o755); err != nil {
				return "", err
			}
		}
		_ = os.Remove(filepath.Join(binDir, name))
		return finalBinary, nil
	}

	return "", fmt.Errorf("downloaded archive did not contain clawdb-server binary")
}

func downloadBinary() (string, error) {
	platform, err := platformID()
	if err != nil {
		return "", err
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	binDir := filepath.Join(home, ".clawdb", "bin")
	if err := os.MkdirAll(binDir, 0o755); err != nil {
		return "", err
	}
	archiveName := fmt.Sprintf("clawdb-server-%s.tar.gz", platform)
	archivePath := filepath.Join(binDir, archiveName)
	for _, base := range releaseBaseCandidates() {
		resp, err := http.Get(fmt.Sprintf("%s/%s", base, archiveName))
		if err != nil {
			continue
		}
		if resp.StatusCode >= 400 {
			_ = resp.Body.Close()
			continue
		}
		file, err := os.Create(archivePath)
		if err != nil {
			_ = resp.Body.Close()
			return "", err
		}
		_, copyErr := io.Copy(file, resp.Body)
		closeErr := file.Close()
		_ = resp.Body.Close()
		if copyErr != nil {
			return "", copyErr
		}
		if closeErr != nil {
			return "", closeErr
		}
		return archivePath, nil
	}

	return "", fmt.Errorf("unable to locate downloadable clawdb-server release archive")
}

func autoProvisionEndpoint(cfg *Config) error {
	if cfg.Endpoint != "" {
		return nil
	}
	if url := os.Getenv("CLAWDB_URL"); url != "" {
		cfg.Endpoint = url
		return nil
	}
	if apiKey := os.Getenv("CLAWDB_API_KEY"); apiKey != "" {
		cfg.APIKey = apiKey
		cfg.Endpoint = "https://cloud.clawdb.dev"
		return nil
	}
	if localServerHealthy() {
		cfg.Endpoint = "http://localhost:50050"
		return nil
	}
	if err := spawnLocalServer("clawdb-server"); err == nil && waitForLocalServer(5*time.Second) {
		cfg.Endpoint = "http://localhost:50050"
		return nil
	}
	if installedBinary, err := installedBinaryPath(); err == nil {
		if _, statErr := os.Stat(installedBinary); statErr == nil {
			if err := spawnLocalServer(installedBinary); err == nil && waitForLocalServer(5*time.Second) {
				cfg.Endpoint = "http://localhost:50050"
				return nil
			}
		}
	}
	if archivePath, err := downloadBinary(); err == nil {
		extractedBinary, extractErr := extractBinary(archivePath)
		if extractErr == nil {
			if err := spawnLocalServer(extractedBinary); err == nil && waitForLocalServer(5*time.Second) {
				cfg.Endpoint = "http://localhost:50050"
				return nil
			}
		}
	}
	if localServerHealthy() {
		cfg.Endpoint = "http://localhost:50050"
		return nil
	}
	return fmt.Errorf("could not auto-provision clawdb-server; start clawdb-server manually or set CLAWDB_URL/CLAWDB_API_KEY")
}
